Improvement/mtime full precision

manuyavuz opened the pull request:
This adds ability to restore mtime values in full nanoseconds precision.

This is required to have reliable incremental builds because some builds systems such as llbuild (https://github.com/apple/swift-llbuild) checks mtime in full precision, rather than checking only seconds.

manuyavuz committed:
Save mtime in full precision

manuyavuz committed:
fixup! Save mtime in full precision

manuyavuz committed:
Use nanoseconds precision

iboB commented:
Thanks for that! That's a really good observation and update. The problem is that it's not backwards compatible. 

Instead of changing the `mtime` key to contain the nanosecond precision it would be a better idea to add an additional optional key `nanoMtime` which contains the improved timestamp. When reading if there is a `nanoMtime` key, it is used to set the file's timestamp, and otherwise we fall back to the `mtime` with second-precision. Thus if a user updates the tool, their old cached timestamps will still be compatible with the new version

iboB commented:
If you don't have time to add the fix, I'll make it this weekend.

manuyavuz committed:
Provide backwards compatibility

Old cache data will include mtime as integer.
This change will make code read old cache, and update with newer
cache policy.

manuyavuz-pointr commented:
Hey @iboB , thanks for reminding backwards compatibility.

I provided compatibility in a different way. Rather than adding a new key, I read available key, and handler different value types (Integer that corresponds to old scheme, vs String which is the new scheme)

I think this made the code shorter. What do you think?

manuyavuz commented:
Hey @iboB , thanks for reminding backwards compatibility.

I provided compatibility in a different way. Rather than adding a new key, I read available key, and handler different value types (Integer that corresponds to old scheme, vs String which is the new scheme)

I think this made the code shorter. What do you think?

iboB commented:
cool. That works
